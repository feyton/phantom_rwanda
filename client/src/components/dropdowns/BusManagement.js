import React, { useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import Swal from 'sweetalert2';
import { axiosBase as axios } from '../../utils/Api.js';
import { ButtonLoading } from '../Button.js';

/* istanbul ignore next */

const BusManagement = ({ row }) => {
  const { plateNumber, busType, seats, driver, route, company, id } =
    row.original;
  const data = {
    plateNumber,
    busType,
    seats,
    driver,
    route,
    company,
    id
  };
  const navigate = useNavigate();
  const [loading, setloading] = useState(false);
  const handleEdit = (busInfo) => {
    navigate(`bus/update/${busInfo.id}`, {
      state: busInfo
    });
  };
  const selectRef = useRef();
  const handleDelete = async (busInfo) => {
    const results = await Swal.fire({
      title: 'Are you sure?',
      html: `Bus <b>${busInfo.plateNumber}</b> will be deleted.`,
      text: 'This driver will be deleted!',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete!'
    });
    if (results.isConfirmed) {
      setloading(true);
      try {
        await axios.delete(`/buses/${busInfo.id}`);
        toast('Bus deleted successfully', { type: 'success' });
        setloading(false);
        selectRef.current.innerHTML = `<p class="mx-auto font-bold text-red font-raleway">Deleted</p>`;
      } catch (error) {
        const errors = error.response.data || error.message;
        toast(errors, { type: 'error' });
      } finally {
        setloading(false);
      }
    }
  };
  const handleAssignRoute = () => {
    navigate(`/dashboard/modal/bus/assign/${plateNumber}`, {
      state: data
    });
  };

  const handleChange = (e) => {
    switch (e.target.value) {
      case 'edit':
        return handleEdit(row.original);
      case 'delete':
        return handleDelete(row.original);
      case 'assign_route':
        return handleAssignRoute(row.original);
    }
  };
  return (
    <div className="z-10" ref={selectRef}>
      {loading ? (
        <ButtonLoading name="Sending" />
      ) : (
        <select
          onChange={handleChange}
          role="select"
          name="busManagement"
          id=""
          className="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        >
          <option hidden>Manage</option>
          <option className="cursor-pointer" value="assign_route">
            Assign Route
          </option>
          <option className="cursor-pointer" value="edit">
            Edit
          </option>
          <option className="cursor-pointer" value="delete">
            Delete
          </option>
        </select>
      )}
    </div>
  );
};

export default BusManagement;
