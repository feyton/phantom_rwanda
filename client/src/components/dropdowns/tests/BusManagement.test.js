import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import BusManagement from '../BusManagement.js';

describe('BusManagement', () => {
  const row = {
    original: {
      busType: 'Coaster',
      plateNumber: 'RAB1234',
      seats: '30',
      driver: null,
      route: null
    }
  };
  it('Should render bus management page', () => {
    const elem = renderer
      .create(
        <BrowserRouter>
          <BusManagement row={row} />
        </BrowserRouter>
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
  it('should edit a bus', async () => {
    const handleEditMock = jest.fn();
    const elem = render(
      <BrowserRouter>
        <BusManagement row={row} handleEdit={handleEditMock} />
      </BrowserRouter>
    );
    fireEvent.change(await screen.findByRole('select'), {
      target: { value: 'edit' }
    });
    expect(elem).toMatchSnapshot();
  });
  it('Should delete a bus', async () => {
    const handleDeleteMock = jest.fn();
    const elem = render(
      <BrowserRouter>
        <BusManagement row={row} handleDelete={handleDeleteMock} />
      </BrowserRouter>
    );
    fireEvent.change(await screen.findByRole('select'), {
      target: { value: 'delete' }
    });
    const r = new Promise((resolve) => setTimeout(resolve, 1000));
    await r;
    expect(elem).toMatchSnapshot();
  });
  it('Should delete a bus', async () => {
    const handleDeleteMock = jest.fn();
    jest.mock('axios');
    const axiosMock = jest.fn(() => Promise.resolve({}));
    const elem = render(
      <BrowserRouter>
        <BusManagement
          row={row}
          handleDelete={handleDeleteMock}
          axios={axiosMock}
        />
      </BrowserRouter>
    );
    fireEvent.change(await screen.findByRole('select'), {
      target: { value: 'delete' }
    });
    expect(elem).toMatchSnapshot();
  });
});
