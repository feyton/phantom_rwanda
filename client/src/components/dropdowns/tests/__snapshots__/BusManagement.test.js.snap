// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusManagement Should delete a bus 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="swal2-shown swal2-height-auto"
  >
    <div
      aria-hidden="true"
    >
      <div
        class="z-10"
      >
        <select
          class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
          id=""
          name="busManagement"
          role="select"
        >
          <option
            hidden=""
          >
            Manage
          </option>
          <option
            class="cursor-pointer"
            value="assign_route"
          >
            Assign Route
          </option>
          <option
            class="cursor-pointer"
            value="edit"
          >
            Edit
          </option>
          <option
            class="cursor-pointer"
            value="delete"
          >
            Delete
          </option>
        </select>
      </div>
    </div>
    <div
      class="swal2-container swal2-center swal2-backdrop-show"
      style="overflow-y: auto;"
    >
      <div
        aria-describedby="swal2-html-container"
        aria-labelledby="swal2-title"
        aria-live="assertive"
        aria-modal="true"
        class="swal2-popup swal2-modal swal2-icon-warning swal2-show"
        role="dialog"
        style="display: grid;"
        tabindex="-1"
      >
        <button
          aria-label="Close this dialog"
          class="swal2-close"
          style="display: none;"
          type="button"
        >
          ×
        </button>
        <ul
          class="swal2-progress-steps"
          style="display: none;"
        />
        <div
          class="swal2-icon swal2-warning swal2-icon-show"
          style="display: flex;"
        >
          <div
            class="swal2-icon-content"
          >
            !
          </div>
        </div>
        <img
          class="swal2-image"
          style="display: none;"
        />
        <h2
          class="swal2-title"
          id="swal2-title"
          style="display: block;"
        >
          Are you sure?
        </h2>
        <div
          class="swal2-html-container"
          id="swal2-html-container"
          style="display: block;"
        >
          Bus 
          <b>
            RAB1234
          </b>
           will be deleted.
        </div>
        <input
          class="swal2-input"
          style="display: none;"
        />
        <input
          class="swal2-file"
          style="display: none;"
          type="file"
        />
        <div
          class="swal2-range"
          style="display: none;"
        >
          <input
            type="range"
          />
          <output />
        </div>
        <select
          class="swal2-select"
          style="display: none;"
        />
        <div
          class="swal2-radio"
          style="display: none;"
        />
        <label
          class="swal2-checkbox"
          for="swal2-checkbox"
          style="display: none;"
        >
          <input
            type="checkbox"
          />
          <span
            class="swal2-label"
          />
        </label>
        <textarea
          class="swal2-textarea"
          style="display: none;"
        />
        <div
          class="swal2-validation-message"
          id="swal2-validation-message"
          style="display: none;"
        />
        <div
          class="swal2-actions"
          style="display: flex;"
        >
          <div
            class="swal2-loader"
          />
          <button
            aria-label=""
            class="swal2-confirm swal2-styled swal2-default-outline"
            style="display: inline-block; background-color: rgb(48, 133, 214);"
            type="button"
          >
            Yes, delete!
          </button>
          <button
            aria-label=""
            class="swal2-deny swal2-styled"
            style="display: none;"
            type="button"
          >
            No
          </button>
          <button
            aria-label=""
            class="swal2-cancel swal2-styled swal2-default-outline"
            style="display: inline-block; background-color: rgb(221, 51, 51);"
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          class="swal2-footer"
          style="display: none;"
        />
        <div
          class="swal2-timer-progress-bar-container"
        >
          <div
            class="swal2-timer-progress-bar"
            style="display: none;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      class="z-10"
    >
      <select
        class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        id=""
        name="busManagement"
        role="select"
      >
        <option
          hidden=""
        >
          Manage
        </option>
        <option
          class="cursor-pointer"
          value="assign_route"
        >
          Assign Route
        </option>
        <option
          class="cursor-pointer"
          value="edit"
        >
          Edit
        </option>
        <option
          class="cursor-pointer"
          value="delete"
        >
          Delete
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BusManagement Should delete a bus 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="swal2-shown swal2-height-auto"
  >
    <div
      aria-hidden="true"
    >
      <div
        class="z-10"
      >
        <select
          class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
          id=""
          name="busManagement"
          role="select"
        >
          <option
            hidden=""
          >
            Manage
          </option>
          <option
            class="cursor-pointer"
            value="assign_route"
          >
            Assign Route
          </option>
          <option
            class="cursor-pointer"
            value="edit"
          >
            Edit
          </option>
          <option
            class="cursor-pointer"
            value="delete"
          >
            Delete
          </option>
        </select>
      </div>
    </div>
    <div
      class="swal2-container swal2-center swal2-backdrop-show"
    >
      <div
        aria-describedby="swal2-html-container"
        aria-labelledby="swal2-title"
        aria-live="assertive"
        aria-modal="true"
        class="swal2-popup swal2-modal swal2-icon-warning"
        role="dialog"
        style="opacity: 0 !important; display: grid;"
        tabindex="-1"
      >
        <button
          aria-label="Close this dialog"
          class="swal2-close"
          style="display: none;"
          type="button"
        >
          ×
        </button>
        <ul
          class="swal2-progress-steps"
          style="display: none;"
        />
        <div
          class="swal2-icon swal2-warning swal2-icon-show"
          style="display: flex;"
        >
          <div
            class="swal2-icon-content"
          >
            !
          </div>
        </div>
        <img
          class="swal2-image"
          style="display: none;"
        />
        <h2
          class="swal2-title"
          id="swal2-title"
          style="display: block;"
        >
          Are you sure?
        </h2>
        <div
          class="swal2-html-container"
          id="swal2-html-container"
          style="display: block;"
        >
          Bus 
          <b>
            RAB1234
          </b>
           will be deleted.
        </div>
        <input
          class="swal2-input"
          style="display: none;"
        />
        <input
          class="swal2-file"
          style="display: none;"
          type="file"
        />
        <div
          class="swal2-range"
          style="display: none;"
        >
          <input
            type="range"
          />
          <output />
        </div>
        <select
          class="swal2-select"
          style="display: none;"
        />
        <div
          class="swal2-radio"
          style="display: none;"
        />
        <label
          class="swal2-checkbox"
          for="swal2-checkbox"
          style="display: none;"
        >
          <input
            type="checkbox"
          />
          <span
            class="swal2-label"
          />
        </label>
        <textarea
          class="swal2-textarea"
          style="display: none;"
        />
        <div
          class="swal2-validation-message"
          id="swal2-validation-message"
          style="display: none;"
        />
        <div
          class="swal2-actions"
          style="display: flex;"
        >
          <div
            class="swal2-loader"
          />
          <button
            aria-label=""
            class="swal2-confirm swal2-styled swal2-default-outline"
            style="display: inline-block; background-color: rgb(48, 133, 214);"
            type="button"
          >
            Yes, delete!
          </button>
          <button
            aria-label=""
            class="swal2-deny swal2-styled"
            style="display: none;"
            type="button"
          >
            No
          </button>
          <button
            aria-label=""
            class="swal2-cancel swal2-styled swal2-default-outline"
            style="display: inline-block; background-color: rgb(221, 51, 51);"
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          class="swal2-footer"
          style="display: none;"
        />
        <div
          class="swal2-timer-progress-bar-container"
        >
          <div
            class="swal2-timer-progress-bar"
            style="display: none;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      class="z-10"
    >
      <select
        class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        id=""
        name="busManagement"
        role="select"
      >
        <option
          hidden=""
        >
          Manage
        </option>
        <option
          class="cursor-pointer"
          value="assign_route"
        >
          Assign Route
        </option>
        <option
          class="cursor-pointer"
          value="edit"
        >
          Edit
        </option>
        <option
          class="cursor-pointer"
          value="delete"
        >
          Delete
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BusManagement Should render bus management page 1`] = `
<div
  className="z-10"
>
  <select
    className="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
    id=""
    name="busManagement"
    onChange={[Function]}
    role="select"
  >
    <option
      hidden={true}
    >
      Manage
    </option>
    <option
      className="cursor-pointer"
      value="assign_route"
    >
      Assign Route
    </option>
    <option
      className="cursor-pointer"
      value="edit"
    >
      Edit
    </option>
    <option
      className="cursor-pointer"
      value="delete"
    >
      Delete
    </option>
  </select>
</div>
`;

exports[`BusManagement should edit a bus 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="z-10"
      >
        <select
          class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
          id=""
          name="busManagement"
          role="select"
        >
          <option
            hidden=""
          >
            Manage
          </option>
          <option
            class="cursor-pointer"
            value="assign_route"
          >
            Assign Route
          </option>
          <option
            class="cursor-pointer"
            value="edit"
          >
            Edit
          </option>
          <option
            class="cursor-pointer"
            value="delete"
          >
            Delete
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="z-10"
    >
      <select
        class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        id=""
        name="busManagement"
        role="select"
      >
        <option
          hidden=""
        >
          Manage
        </option>
        <option
          class="cursor-pointer"
          value="assign_route"
        >
          Assign Route
        </option>
        <option
          class="cursor-pointer"
          value="edit"
        >
          Edit
        </option>
        <option
          class="cursor-pointer"
          value="delete"
        >
          Delete
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
